datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -1685,8 +1725,33 @@ pub struct Extension {
     pub node: Arc<dyn UserDefinedLogicalNode>,
 }
 
+struct ExtensionExplainDisplay<'a> {
+    extension: &'a Extension,
+}
+
+impl Display for ExtensionExplainDisplay<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        self.extension.node.fmt_for_explain(f)
+    }
+}
+
+impl PartialEq for Extension {
+    fn eq(&self, other: &Self) -> bool {
+        format!("{}", ExtensionExplainDisplay { extension: self })
+            == format!("{}", ExtensionExplainDisplay { extension: other })
+    }
+}
+
+impl Eq for Extension {}
+
+impl Hash for Extension {
```

Comment:
I am not sure about using the textual output for equality comparison as someone who has implemented Extension may have a constant output, for example. I think a safer (though backwards incompatible change) would be to make `UserDefinedLogicalNode` also be `Hash` and `PartialEq`

Like

```rust
trait UserDefinedLogicalNode: Hash + PartialEq {
...
}
```

Reply:
Couldn't add `Hash + PartialEq`, because `UserDefinedLogicalNode` must be *object-safe*.

Instead added `dyn_eq` and `dyn_hash` methods serving the same purpose.